dsl/src/commonMain/kotlin/it/unibo/collektive/aggregate/api/impl/AggregateContext.kt

```diff
@@ -0,0 +1,135 @@
+ * and the [previousState] of the device.
+ */
+internal class AggregateContext(
+    override val localId: ID,
+    private val messages: Iterable<InboundMessage>,
+    private val previousState: State,
+) : Aggregate {
+
+    private val stack = Stack()
+    private var state: State = mapOf()
+    private var toBeSent = OutboundMessage(localId, emptyMap())
+
+    /**
+     * Messages to send to the other nodes.
+     */
+    fun messagesToSend(): OutboundMessage = toBeSent
+
+    /**
+     * Return the current state of the device as a new state.
+     */
+    fun newState(): State = state
+
+    private fun <T> newField(localValue: T, others: Map<ID, T>): Field<T> = Field(localId, localValue, others)
+
+    override fun <X> exchange(initial: X, body: (Field<X>) -> Field<X>): Field<X> {
+        val messages = messagesAt<X>(stack.currentPath())
+        val previous = stateAt(stack.currentPath(), initial)
+        val subject = newField(previous, messages)
+        return body(subject).also { field ->
+            val message = SingleOutboundMessage(field.localValue, field.excludeSelf())
+            val path = stack.currentPath()
+            check(!toBeSent.messages.containsKey(path)) {
+                """
+                    Alignment was broken by multiple aligned calls with the same path: $path.
+                    The most likely cause is an aggregate function call within a loop
+                """.trimIndent()
+            }
+            toBeSent = toBeSent.copy(messages = toBeSent.messages + (stack.currentPath() to message))
+            state = state + (stack.currentPath() to field.localValue)
+        }
+    }
+
+    override fun <Init, Ret> exchanging(
+        initial: Init,
+        body: YieldingContext<Field<Init>, Field<Ret>>.(Field<Init>) -> YieldingResult<Field<Init>, Field<Ret>>,
+    ): Field<Ret> {
+        val messages = messagesAt<Init>(stack.currentPath())
+        val previous = stateAt(stack.currentPath(), initial)
+        val subject = newField(previous, messages)
+        val context = YieldingContext<Field<Init>, Field<Ret>>()
+        var res: Option<YieldingResult<Field<Init>, Field<Ret>>>
+        body(context, subject).also {
+            res = it.some()
+            state = state + (stack.currentPath() to it.toReturn.localValue)
+        }
+        return res.getOrElse { error("This error should never be thrown") }.toReturn
+    }
+
+    override fun <Initial, Return> repeating(
+        initial: Initial,
+        transform: YieldingContext<Initial, Return>.(Initial) -> YieldingResult<Initial, Return>,
+    ): Return {
+        val context = YieldingContext<Initial, Return>()
+        var res: Option<YieldingResult<Initial, Return>>
+        transform(context, stateAt(stack.currentPath(), initial)).also {
+            res = it.some()
+            state = state + (stack.currentPath() to it.toReturn)
+        }
+        return res.getOrElse { error("This error should never be thrown") }.toReturn
+    }
```

Comment:
Can't we implement the yielding form correctly and then implement the non-yielding form by calling it?

Also I believe we can `inline` most of the implementation HOF functions